lib/elixir/lib/string_io.ex

```diff
@@ -23,7 +23,62 @@ defmodule StringIO do
 
   If the `:capture_prompt` option is set to `true`,
   prompts (specified as arguments to `IO.get*` functions)
-  are captured.
+  are captured in the output.
+
+  The device will be created and sent to the function given.
+  When the function returns, the device will be closed. The final
+  result will be a tuple with `:ok` and the result of the function.
+
+  ## Examples
+
+    iex> {:ok, _contents} = StringIO.open("foo", [], fn(pid) ->
+    ...>    input = IO.gets(pid, ">")
+    ...>    IO.write(pid, "The input was \#{input}")
+    ...>    StringIO.contents(pid)
+    ...> end)
+    {:ok, {"", "The input was foo"}}
+
+    iex> {:ok, _contents} = StringIO.open("foo", [capture_prompt: true], fn(pid) ->
+    ...>    input = IO.gets(pid, ">")
+    ...>    IO.write(pid, "The input was \#{input}")
+    ...>    StringIO.contents(pid)
+    ...> end)
+    {:ok, {"", ">The input was foo"}}
+
+  """
+  @spec open(binary, keyword, (pid -> res)) :: {:ok, res} when res: var
```

Comment:
The spec says it returns `{:ok, res}` but the code says it may return the error from GenServer.start_link. Maybe we also don't need a case and instead we should do `{:ok, pid} = GenServer...`?

Reply:
I patterned my implementation after the implementation of `File.open` (https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/file.ex#L1297) which does use a case statement.  The spec there also calls out the fact that the result may be `{:error, posix}`.  
If you agree that matching the behavior of that function is the right approach, I can revise the `@spec`.

The alternative, I suppose, is for the function to raise a match error if the `GenServer` can't start up.  That would make this an `open!` function, wouldn't it?

Reply:
We would use `{:ok, pid}` because there is no chance it can fail. If it can fail, then we should return `{:error, _}` and update the `@spec` but i don't think it can ever happen. Hence pattern matching. :)

Reply:
Ok.  I'll remove the case and add the pattern match.  Thanks!

Reply:
In a completely unrelated note, I also noticed some tests fail (on Mac OS X anyway) if there is a space in the path leading up to where you've cloned the elixir repo.